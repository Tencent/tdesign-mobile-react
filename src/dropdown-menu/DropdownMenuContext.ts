import noop from 'lodash/noop';
import React from 'react';
import { dropdownMenuDefaultProps } from './defaultProps';

const DropdownMenuContext = React.createContext({
  ...dropdownMenuDefaultProps,
  activedId: '',
  onChangeActivedId: noop,
});

export default DropdownMenuContext;
