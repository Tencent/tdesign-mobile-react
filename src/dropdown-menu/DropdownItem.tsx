import { useClickAway } from 'ahooks';
import cx from 'classnames';
import uniqueId from 'lodash/uniqueId';
import React, { FC, useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react';
import { CaretDownSmallIcon, CaretUpSmallIcon } from 'tdesign-icons-react';
import { Button, Checkbox, Popup, Radio, RadioGroup } from 'tdesign-mobile-react';
import useDefault from 'tdesign-mobile-react/_util/useDefault';
import CheckboxGroup from 'tdesign-mobile-react/checkbox/CheckboxGroup';
import { StyledProps } from 'tdesign-mobile-react/common';
import useDefaultProps from 'tdesign-mobile-react/hooks/useDefaultProps';
import useConfig from '../_util/useConfig';
import { dropdownItemDefaultProps } from './defaultProps';
import DropdownMenuContext from './DropdownMenuContext';
import type { TdDropdownItemProps } from './type';

export interface DropdownItemProps extends TdDropdownItemProps, StyledProps {}

const DropdownItem: FC<DropdownItemProps> = (props) => {
  const {
    className,
    style,
    disabled,
    options: inputOptions,
    optionsColumns,
    label,
    value,
    defaultValue,
    onChange,
    multiple,
    onConfirm,
    onReset,
    footer,
    keys,
  } = useDefaultProps<DropdownItemProps>(props, dropdownItemDefaultProps);
  const { classPrefix } = useConfig();
  const itemClass = `${classPrefix}-dropdown-item`;
  const [innerValue, setInnerValue] = useDefault(value, defaultValue, onChange);
  const [modalValue, setModalValue] = useState(innerValue);

  const options = useMemo(
    () =>
      inputOptions.map((item) => ({
        value: item[keys?.value ?? 'value'],
        label: item[keys?.label ?? 'label'],
        disabled: item[keys?.disabled ?? 'disabled'],
      })),
    [keys, inputOptions],
  );

  const [id] = useState(uniqueId());

  const { direction, activedId, onChangeActivedId, showOverlay, zIndex, closeOnClickOverlay } =
    useContext(DropdownMenuContext);

  const labelText = useMemo(
    () => label || options.find((item) => item.value === innerValue)?.label || '',
    [options, label, innerValue],
  );

  const isActived = id === activedId;

  const menuItemRef = useRef<HTMLDivElement>();
  const itemRef = useRef<HTMLDivElement>();

  const getDropdownItemStyle = () => {
    const ele = menuItemRef.current;
    if (!ele) {
      return {};
    }

    const { top, bottom } = ele.getBoundingClientRect();

    if (direction === 'up') {
      return {
        transform: 'rotateX(180deg) rotateY(180deg)',
        zIndex,
        bottom: `calc(100vh - ${top}px)`,
      };
    }

    return {
      zIndex,
      top: `${bottom}px`,
    };
  };

  useClickAway(() => {
    if (!isActived || !closeOnClickOverlay) {
      return;
    }
    onChangeActivedId('');
  }, itemRef);

  useEffect(() => {
    if (isActived) {
      setModalValue(innerValue);
    }
  }, [isActived, innerValue]);

  const attach = useCallback(() => itemRef.current || document.body, []);

  return (
    <>
      <div
        ref={menuItemRef}
        className={cx(`${classPrefix}-dropdown-menu__item`, {
          [`${classPrefix}-dropdown-menu__item--active`]: isActived,
          [`${classPrefix}-dropdown-menu__item--disabled`]: disabled,
        })}
        onClick={(e) => {
          if (disabled) {
            return;
          }
          onChangeActivedId(isActived ? '' : id);
          if (!isActived) {
            e.stopPropagation();
          }
        }}
      >
        <div className={`${classPrefix}-dropdown-menu__title`}>{labelText}</div>
        {direction === 'down' ? (
          <CaretDownSmallIcon
            className={cx(`${classPrefix}-dropdown-menu__icon`, {
              [`${classPrefix}-dropdown-menu__icon--active`]: isActived,
            })}
          />
        ) : (
          <CaretUpSmallIcon
            className={cx(`${classPrefix}-dropdown-menu__icon`, {
              [`${classPrefix}-dropdown-menu__icon--active`]: isActived,
            })}
          />
        )}
      </div>
      {isActived ? (
        <div
          key={id}
          className={cx(itemClass, className)}
          style={{
            ...style,
            ...getDropdownItemStyle(),
          }}
          ref={itemRef}
        >
          {/* TODO Popup 暂不支持 duration */}
          <Popup
            attach={attach}
            visible={isActived}
            onVisibleChange={(visible) => {
              if (!visible) {
                onChangeActivedId('');
              }
            }}
            closeOnOverlayClick={closeOnClickOverlay}
            showOverlay={showOverlay}
            zIndex={zIndex}
            style={{
              position: 'absolute',
              borderRadius: 0,
            }}
            overlayProps={{
              style: {
                position: 'absolute',
              },
            }}
            className={`${itemClass}__popup-host`}
          >
            <div className={cx(`${itemClass}__content`, `${classPrefix}-popup__content`)}>
              <div
                className={cx(`${itemClass}__body`)}
                style={
                  direction === 'up'
                    ? {
                        transform: 'rotateX(180deg) rotateY(180deg)',
                      }
                    : {}
                }
              >
                {props.children ? (
                  props.children
                ) : (
                  <>
                    {/* TODO checkbox 组件未升级 样式不对 */}
                    {multiple ? (
                      <CheckboxGroup
                        value={modalValue as (string | number)[]}
                        onChange={(value) => {
                          setModalValue(value);
                        }}
                        // className={`itemClass__checkbox-group`}
                        style={{
                          gridTemplateColumns: `repeat(${optionsColumns}, 1fr)`,
                        }}
                      >
                        {options.map((item, index) => (
                          <Checkbox
                            key={index}
                            value={item.value as string | number}
                            label={item.label}
                            disabled={item.disabled}
                          />
                        ))}
                      </CheckboxGroup>
                    ) : (
                      <RadioGroup
                        value={modalValue as string | number}
                        onChange={(value: string | number) => {
                          setModalValue(value);
                          setInnerValue(value);
                          onChangeActivedId('');
                        }}
                      >
                        {/* TODO radio 暂不支持 icon line */}
                        {options.map((item, index) => (
                          <Radio
                            key={index}
                            value={item.value as string | number}
                            label={item.label}
                            disabled={item.disabled}
                          />
                        ))}
                      </RadioGroup>
                    )}
                  </>
                )}
              </div>

              {footer ? footer : null}

              {multiple && !footer ? (
                <div className={`${itemClass}__footer`}>
                  <Button
                    disabled={Array.isArray(modalValue) && modalValue.length === 0}
                    theme="light"
                    className={`${itemClass}__footer-btn ${itemClass}__reset-btn`}
                    onClick={() => {
                      if (typeof onReset === 'function') {
                        onReset(modalValue);
                      } else {
                        setModalValue(innerValue);
                      }
                    }}
                  >
                    重置
                  </Button>
                  <Button
                    disabled={Array.isArray(modalValue) && modalValue.length === 0}
                    theme="primary"
                    className={`${itemClass}__footer-btn ${itemClass}__confirm-btn`}
                    onClick={() => {
                      if (typeof onConfirm === 'function') {
                        onConfirm(modalValue);
                      } else {
                        setInnerValue(modalValue);
                      }
                      onChangeActivedId('');
                    }}
                  >
                    确定
                  </Button>
                </div>
              ) : null}
            </div>
          </Popup>
        </div>
      ) : null}
    </>
  );
};

DropdownItem.displayName = 'DropdownItem';

export default DropdownItem;
