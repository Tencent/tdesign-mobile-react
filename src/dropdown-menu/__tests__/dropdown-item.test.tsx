import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { vi, describe, it, expect, afterEach } from 'vitest';
import DropdownItem from '../DropdownItem';
import DropdownMenuContext from '../DropdownMenuContext';

vi.mock('lodash-es/uniqueId', () => ({ default: () => 'fixed-id' }));

describe('DropdownItem', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  it('returns empty style when menuItemRef is null', () => {
    const utils = render(
      <DropdownMenuContext.Provider
        value={{
          activedId: '',
          direction: 'down',
          onChangeActivedId: () => {},
          showOverlay: false,
          zIndex: 10,
          closeOnClickOverlay: false,
        }}
      >
        <DropdownItem options={[{ value: '1', label: 'one', disabled: false } as any]} />
      </DropdownMenuContext.Provider>,
    );

    expect(utils.container).toBeTruthy();
  });

  it('renders popup when active and handles direction up/down in style', async () => {
    const onChangeActivedId = vi.fn();

    const { container, getByText } = render(
      <DropdownMenuContext.Provider
        value={{
          activedId: 'fixed-id',
          direction: 'up',
          onChangeActivedId,
          showOverlay: true,
          zIndex: 99,
          closeOnClickOverlay: true,
        }}
      >
        <DropdownItem options={[{ value: 'a', label: 'A', disabled: false } as any]} multiple={false} />
      </DropdownMenuContext.Provider>,
    );

    expect(getByText('A')).toBeTruthy();

    await waitFor(() => {
      const item = container.querySelector('[class*="dropdown-item"]');
      expect(item).toBeTruthy();
    });
  });

  it('calculates style when menuItemRef has bounding rect and direction up', async () => {
    const onChangeActivedId = vi.fn();

    const { container, rerender } = render(
      <DropdownMenuContext.Provider
        value={{
          activedId: '',
          direction: 'up',
          onChangeActivedId,
          showOverlay: true,
          zIndex: 77,
          closeOnClickOverlay: true,
        }}
      >
        <DropdownItem options={[{ value: 'x', label: 'X', disabled: false } as any]} />
      </DropdownMenuContext.Provider>,
    );

    const menuTrigger = container.querySelector('.t-dropdown-menu__item') as HTMLElement | null;
    if (menuTrigger) {
      menuTrigger.getBoundingClientRect = () => ({ top: 10, bottom: 50 }) as DOMRect;
    }

    rerender(
      <DropdownMenuContext.Provider
        value={{
          activedId: 'fixed-id',
          direction: 'up',
          onChangeActivedId,
          showOverlay: true,
          zIndex: 77,
          closeOnClickOverlay: true,
        }}
      >
        <DropdownItem options={[{ value: 'x', label: 'X', disabled: false } as any]} />
      </DropdownMenuContext.Provider>,
    );

    await waitFor(() => {
      const wrapper = container.querySelector('.t-dropdown-item') as HTMLElement | null;
      expect(wrapper).toBeTruthy();
      if (wrapper) {
        const { bottom } = wrapper.style;
        const { top } = wrapper.style;
        expect(bottom || top).toBeTruthy();
      }
    });
  });

  it('calls onReset when provided and falls back when not provided', async () => {
    const onReset = vi.fn();
    const onConfirm = vi.fn();
    const onChangeActivedId = vi.fn();

    const { getByText, rerender } = render(
      <DropdownMenuContext.Provider
        value={{
          activedId: 'fixed-id',
          direction: 'down',
          onChangeActivedId,
          showOverlay: true,
          zIndex: 1,
          closeOnClickOverlay: true,
        }}
      >
        <DropdownItem
          options={[{ value: '1', label: 'one', disabled: false } as any]}
          multiple
          onReset={onReset}
          onConfirm={onConfirm}
        />
      </DropdownMenuContext.Provider>,
    );

    const resetBtn = getByText('重置');
    const confirmBtn = getByText('确定');
    fireEvent.click(resetBtn);
    fireEvent.click(confirmBtn);

    expect(onReset).toHaveBeenCalled();
    expect(onConfirm).toHaveBeenCalled();

    rerender(
      <DropdownMenuContext.Provider
        value={{
          activedId: 'fixed-id',
          direction: 'down',
          onChangeActivedId,
          showOverlay: true,
          zIndex: 1,
          closeOnClickOverlay: true,
        }}
      >
        <DropdownItem options={[{ value: '1', label: 'one', disabled: false } as any]} multiple />
      </DropdownMenuContext.Provider>,
    );

    fireEvent.click(getByText('重置'));
    fireEvent.click(getByText('确定'));
  });

  it('supports custom keys mapping for options', () => {
    const onChangeActivedId = vi.fn();
    const { getByText } = render(
      <DropdownMenuContext.Provider
        value={{
          activedId: 'fixed-id',
          direction: 'down',
          onChangeActivedId,
          showOverlay: false,
          zIndex: 1,
          closeOnClickOverlay: false,
        }}
      >
        <DropdownItem
          options={[{ k: '1', v: 'one', d: false } as any]}
          keys={{ value: 'k', label: 'v', disabled: 'd' }}
        />
      </DropdownMenuContext.Provider>,
    );

    expect(getByText('one')).toBeTruthy();
  });

  it('renders children and footer when provided', () => {
    const onChangeActivedId = vi.fn();
    const { getByText } = render(
      <DropdownMenuContext.Provider
        value={{
          activedId: 'fixed-id',
          direction: 'down',
          onChangeActivedId,
          showOverlay: true,
          zIndex: 1,
          closeOnClickOverlay: true,
        }}
      >
        <DropdownItem options={[{ value: '1', label: 'one', disabled: false } as any]}>
          <div>custom-body</div>
        </DropdownItem>
      </DropdownMenuContext.Provider>,
    );

    expect(getByText('custom-body')).toBeTruthy();
  });

  it('multiple mode: buttons disabled when modalValue empty then enabled after change', async () => {
    const onChangeActivedId = vi.fn();
    const { getByText, container } = render(
      <DropdownMenuContext.Provider
        value={{
          activedId: 'fixed-id',
          direction: 'down',
          onChangeActivedId,
          showOverlay: true,
          zIndex: 1,
          closeOnClickOverlay: true,
        }}
      >
        <DropdownItem options={[{ value: '1', label: 'one', disabled: false } as any]} multiple />
      </DropdownMenuContext.Provider>,
    );

    const resetBtn = getByText('重置');
    const confirmBtn = getByText('确定');
    expect(resetBtn).toBeTruthy();
    expect(confirmBtn).toBeTruthy();

    const input = container.querySelector('input[type="radio"]') as HTMLInputElement | null;
    if (input) {
      fireEvent.click(input);
    }

    expect(getByText('重置')).toBeTruthy();
  });

  it('does not activate when disabled', () => {
    const onChangeActivedId = vi.fn();
    const { container } = render(
      <DropdownMenuContext.Provider
        value={{
          activedId: '',
          direction: 'down',
          onChangeActivedId,
          showOverlay: false,
          zIndex: 1,
          closeOnClickOverlay: false,
        }}
      >
        <DropdownItem options={[{ value: '1', label: 'one', disabled: false } as any]} disabled />
      </DropdownMenuContext.Provider>,
    );

    const trigger = container.querySelector('.t-dropdown-menu__item') as HTMLElement | null;
    if (trigger) {
      fireEvent.click(trigger);
    }

    expect(onChangeActivedId).not.toHaveBeenCalled();
  });

  it('calls onChangeActivedId when clicking outside while active and closeOnClickOverlay true', async () => {
    const onChangeActivedId = vi.fn();
    render(
      <DropdownMenuContext.Provider
        value={{
          activedId: 'fixed-id',
          direction: 'down',
          onChangeActivedId,
          showOverlay: true,
          zIndex: 1,
          closeOnClickOverlay: true,
        }}
      >
        <DropdownItem options={[{ value: '1', label: 'one', disabled: false } as any]} />
      </DropdownMenuContext.Provider>,
    );

    const overlay = document.querySelector('.t-overlay') as HTMLElement | null;
    if (overlay) {
      fireEvent.click(overlay);
      fireEvent.mouseDown(overlay);
    } else {
      fireEvent.mouseDown(document.body);
    }

    await waitFor(() => {
      expect(onChangeActivedId).toHaveBeenCalled();
    });
  });

  it('renders custom footer when footer prop provided', () => {
    const onChangeActivedId = vi.fn();
    const { getByText } = render(
      <DropdownMenuContext.Provider
        value={{
          activedId: 'fixed-id',
          direction: 'down',
          onChangeActivedId,
          showOverlay: true,
          zIndex: 1,
          closeOnClickOverlay: true,
        }}
      >
        <DropdownItem
          options={[{ value: '1', label: 'one', disabled: false } as any]}
          footer={<div>my-footer</div>}
          multiple
        />
      </DropdownMenuContext.Provider>,
    );

    expect(getByText('my-footer')).toBeTruthy();
  });
});
