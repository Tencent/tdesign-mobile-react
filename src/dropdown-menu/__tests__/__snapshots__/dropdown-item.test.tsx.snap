// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropdownItem > render / props > calculates style when menuItemRef has bounding rect and direction up 1`] = `
<div>
  <div
    class="t-dropdown-menu__item t-dropdown-menu__item--active"
  >
    <div
      class="t-dropdown-menu__title"
    />
    <svg
      class="t-icon t-icon-caret-up-small t-dropdown-menu__icon t-dropdown-menu__icon--active"
      fill="none"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M12.0001 9L17.5 14.5L6.5 14.5L12.0001 9Z"
        fill="currentColor"
      />
    </svg>
  </div>
  <div
    class="t-dropdown-item"
    style="z-index: 77; bottom: calc(100vh - 10px);"
  >
    <div
      class="t-portal-wrapper"
    >
      <div
        class="t-overlay t-overlay-fade-appear t-overlay-fade-appear-active"
        style="position: absolute; z-index: 1000; animation-duration: 240ms;"
      />
      <div
        class="t-popup t-popup--bottom"
        style="z-index: 77; animation-fill-mode: forwards; position: absolute; overflow: hidden;"
      >
        <div
          class="t-dropdown-item__content t-popup__content"
        >
          <div
            class="t-dropdown-item__body"
          >
            <div
              class="t-radio-group t-dropdown-item__radio-group"
            >
              <div
                class="t-radio t-radio--left t-radio--block"
              >
                <input
                  class="t-radio__original"
                  type="radio"
                  value="x"
                />
                <div
                  class="t-radio__icon t-radio__icon--left"
                >
                  <div
                    class="placeholder"
                  />
                </div>
                <div
                  class="t-radio__content"
                >
                  <span
                    class="t-radio__title"
                    style="-webkit-line-clamp: 3;"
                  >
                    X
                  </span>
                </div>
                <div
                  class="t-radio__border t-radio__border--left"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropdownItem > render / props > renders children and footer when provided 1`] = `
<body
  class="t-overlay--lock"
>
  <div>
    <div
      class="t-dropdown-menu__item t-dropdown-menu__item--active"
    >
      <div
        class="t-dropdown-menu__title"
      />
      <svg
        class="t-icon t-icon-caret-down-small t-dropdown-menu__icon t-dropdown-menu__icon--active"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M12.0001 15L17.5 9.5H6.5L12.0001 15Z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="t-dropdown-item"
    >
      <div
        class="t-portal-wrapper"
      >
        <div
          class="t-overlay t-overlay-fade-appear t-overlay-fade-appear-active"
          style="position: absolute; z-index: 1000; animation-duration: 240ms;"
        />
        <div
          class="t-popup t-popup--top"
          style="z-index: 1; animation-fill-mode: forwards; position: absolute; overflow: hidden;"
        >
          <div
            class="t-dropdown-item__content t-popup__content"
          >
            <div
              class="t-dropdown-item__body"
            >
              <div>
                custom-body
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`DropdownItem > render / props > renders custom footer when footer prop provided 1`] = `
<body
  class="t-overlay--lock"
>
  <div>
    <div
      class="t-dropdown-menu__item t-dropdown-menu__item--active"
    >
      <div
        class="t-dropdown-menu__title"
      />
      <svg
        class="t-icon t-icon-caret-down-small t-dropdown-menu__icon t-dropdown-menu__icon--active"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M12.0001 15L17.5 9.5H6.5L12.0001 15Z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="t-dropdown-item"
    >
      <div
        class="t-portal-wrapper"
      >
        <div
          class="t-overlay t-overlay-fade-appear t-overlay-fade-appear-active"
          style="position: absolute; z-index: 1000; animation-duration: 240ms;"
        />
        <div
          class="t-popup t-popup--top"
          style="z-index: 1; animation-fill-mode: forwards; position: absolute; overflow: hidden;"
        >
          <div
            class="t-dropdown-item__content t-popup__content"
          >
            <div
              class="t-dropdown-item__body"
            >
              <div
                class="t-checkbox-group t-dropdown-item__checkbox-group"
                style="grid-template-columns: repeat(1, 1fr);"
              >
                <div
                  class="t-checkbox t-checkbox--left t-checkbox--block t-dropdown-item__checkbox-item t-checkbox--tag"
                >
                  <div
                    class="t-checkbox__content"
                  >
                    <div
                      class="t-checkbox__title"
                      style="-webkit-line-clamp: 3;"
                    >
                      one
                    </div>
                    <div
                      class="t-checkbox__description"
                      style="-webkit-line-clamp: 5;"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div>
              my-footer
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`DropdownItem > render / props > renders popup when active and handles direction up/down in style 1`] = `
<div>
  <div
    class="t-dropdown-menu__item t-dropdown-menu__item--active"
  >
    <div
      class="t-dropdown-menu__title"
    />
    <svg
      class="t-icon t-icon-caret-up-small t-dropdown-menu__icon t-dropdown-menu__icon--active"
      fill="none"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M12.0001 9L17.5 14.5L6.5 14.5L12.0001 9Z"
        fill="currentColor"
      />
    </svg>
  </div>
  <div
    class="t-dropdown-item"
  >
    <div
      class="t-portal-wrapper"
    >
      <div
        class="t-overlay t-overlay-fade-appear t-overlay-fade-appear-active"
        style="position: absolute; z-index: 1000; animation-duration: 240ms;"
      />
      <div
        class="t-popup t-popup--bottom"
        style="z-index: 99; animation-fill-mode: forwards; position: absolute; overflow: hidden;"
      >
        <div
          class="t-dropdown-item__content t-popup__content"
        >
          <div
            class="t-dropdown-item__body"
          >
            <div
              class="t-radio-group t-dropdown-item__radio-group"
            >
              <div
                class="t-radio t-radio--left t-radio--block"
              >
                <input
                  class="t-radio__original"
                  type="radio"
                  value="a"
                />
                <div
                  class="t-radio__icon t-radio__icon--left"
                >
                  <div
                    class="placeholder"
                  />
                </div>
                <div
                  class="t-radio__content"
                >
                  <span
                    class="t-radio__title"
                    style="-webkit-line-clamp: 3;"
                  >
                    A
                  </span>
                </div>
                <div
                  class="t-radio__border t-radio__border--left"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropdownItem > render / props > returns empty style when menuItemRef is null 1`] = `
<DocumentFragment>
  <div
    class="t-dropdown-menu__item"
  >
    <div
      class="t-dropdown-menu__title"
    />
    <svg
      class="t-icon t-icon-caret-down-small t-dropdown-menu__icon"
      fill="none"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M12.0001 15L17.5 9.5H6.5L12.0001 15Z"
        fill="currentColor"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`DropdownItem > render / props > supports custom keys mapping for options 1`] = `
<body
  class=""
>
  <div>
    <div
      class="t-dropdown-menu__item t-dropdown-menu__item--active"
    >
      <div
        class="t-dropdown-menu__title"
      />
      <svg
        class="t-icon t-icon-caret-down-small t-dropdown-menu__icon t-dropdown-menu__icon--active"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M12.0001 15L17.5 9.5H6.5L12.0001 15Z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="t-dropdown-item"
    >
      <div
        class="t-portal-wrapper"
      >
        <div
          class="t-popup t-popup--top"
          style="z-index: 1; animation-fill-mode: forwards; position: absolute; overflow: hidden;"
        >
          <div
            class="t-dropdown-item__content t-popup__content"
          >
            <div
              class="t-dropdown-item__body"
            >
              <div
                class="t-radio-group t-dropdown-item__radio-group"
              >
                <div
                  class="t-radio t-radio--left t-radio--block"
                >
                  <input
                    class="t-radio__original"
                    type="radio"
                    value="1"
                  />
                  <div
                    class="t-radio__icon t-radio__icon--left"
                  >
                    <div
                      class="placeholder"
                    />
                  </div>
                  <div
                    class="t-radio__content"
                  >
                    <span
                      class="t-radio__title"
                      style="-webkit-line-clamp: 3;"
                    >
                      one
                    </span>
                  </div>
                  <div
                    class="t-radio__border t-radio__border--left"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
